### PR TITLE
Handle blockstore insert dup checks

### PR DESCRIPTION
**Pull request by carllin**
#### Problem
should_insert_data_shred() should check just_inserted_data_shreds() for certain shreds

#### Summary of Changes
Add `get_data_shred_from_just_inserted_or_db()` to fetch data shreds safely

Fixes #

### CODE REVIEW

File: ledger/src/blockstore.rs [hunk 1]
```diff
@@ -1262,82 +1264,110 @@ impl Blockstore {
...
+                        self.get_data_shred(slot, last_index)
+                            .unwrap()
+                            .map(Cow::Owned)
+                    };
+
+                if ending_shred.is_none() {
+                    // The slot data might have been cleared between us fetching the SlotMeta and
+                    // retrieving the `ending_shred`. Slot data might have been cleared by
+                    // `Blockstore::clear_unconfirmed_slot()` due to events like duplicate block detection.
+                    // In this case, try to get the updated SlotMeta and try again
```
Comment (jstarry) [inline ledger/src/blockstore.rs:1303]: Looks like this is going to need a call to `get_slot_meta_entry` or something similar to actually get updated `SlotMeta`, right? Looping again isn't going to change anything

Comment (jstarry) [reply ledger/src/blockstore.rs:1303]: We should probably have an exit hatch for if `SlotMeta` ever gets corrupted such that its `last_index` is still set but the shred doesn't exist

Comment (jstarry) [reply ledger/src/blockstore.rs:1303]: Maybe for now we can just have this return an error? This might actually be an invalid shred

Comment (carllin) [reply ledger/src/blockstore.rs]: ha shoot forgot to purge this, it's no longer necessary

**Review by jstarry — DISMISSED** [review]

**Review by sakridge — DISMISSED** [review]
lgtm

**Comment by carllin** [comment]
Ah dang, there's an annoying case here for duplicate blocks as well.

If somebody sends a batch with two shreds both last_index set to true, say for two shred indexes in the order `10, 9`. Then the shred_index `9` won't be inserted because it fails the should_insert check. If shred `9` is part of the confirmed version of this slot, then this slot will never finish replay, and moreover, the slot is not marked dead because it's never replayed because the entries are never successfully deserialized (blockstore just sees a hole for shred `9` and is waiting for a consecutive set of shreds to deserialize)

**Pull request merged**